File: examples/haskell/pink-fork-linux.hs
-- A simple example showing how to do tracing fork on Linux

import Data.Maybe
import Control.Monad
import System.Exit
import System.Posix.Process (forkProcess)
import System.Posix.Signals
import System.PinkTrace.Event
import System.PinkTrace.Trace

main :: IO ()
main = do
    pid <- forkProcess child
    -- Note this 'getProcessStatus' function is from the System.PinkTrace.Event
    -- module. It's a simple re-implementation of 'getProcessStatus' function
    -- of the System.Posix.Process module which gives information about ptrace
    -- events.
    status <- getProcessStatus True False pid
    case fromJust status of
        Exited (ExitFailure ret) -> putStrLn ("Child exited with code: " ++ (show ret)) >> exitFailure
        Exited ExitSuccess       -> putStrLn ("Child exited with code: 0") >> exitFailure
        Terminated sig           -> putStrLn ("Child was terminated: " ++ (show sig)) >> exitFailure
        StoppedTrace ev          -> putStrLn ("Child has done an unexpected ptrace event " ++ (show ev)) >> exitFailure
        Stopped sig              -> unless (sig == softwareStop)
                                        (putStrLn ("Wrong signal: " ++ show sig) >> exitFailure)

    -- Setup the child
    traceSetup opt pid

    -- Nothing interesting in this example
    -- Let the child continue its execution.
    traceContinue 1 nullSignal pid
    where
        opt :: TraceOption
        opt = TraceOption { traceOptionSysGood   = True
                          , traceOptionFork      = False
                          , traceOptionVFork     = False
                          , traceOptionClone     = False
                          , traceOptionExec      = False
                          , traceOptionVForkDone = False
                          , traceOptionExit      = False
                          }
        child :: IO ()
        child = do
            traceMe
            raiseSignal softwareStop
            putStrLn "hello world"

File: examples/c/pink-fork-linux.c
/* vim: set cino= fo=croql sw=8 ts=8 sts=0 noet cin fdm=syntax : */

/**
 * \file
 *
 * Example \ref pink-fork-linux.c "pink-fork-linux.c" .
 **/

/**
 * \example pink-fork-linux.c
 *
 * Example demonstrating how to fork and start tracing a process on Linux.
 **/

#include <assert.h>
#include <errno.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/wait.h>

#include <pinktrace/pink.h>

int
main(void)
{
	int status;
	pid_t pid;
	pink_event_t event;

	/* Fork and start tracing. */
	if ((pid = fork()) < 0) {
		perror("fork");
		return EXIT_FAILURE;
	}
	else if (!pid) {
		/* Set up for tracing. */
		if (!pink_trace_me()) {
			perror("pink_trace_me");
			_exit(EXIT_FAILURE);
		}
		/* Stop, so that the parent has the chance to set up options
		 * for tracing. */
		kill(getpid(), SIGSTOP);

		/* Here's the point where you can do whatever the traced child
		 * should do. */
		printf("hello world\n");

		_exit(EXIT_SUCCESS);
	}
	else {
		waitpid(pid, &status, 0);
		event = pink_event_decide(status);
		assert(event == PINK_EVENT_STOP);

		/* Set up the child for tracing */
		if (!pink_trace_setup(pid, PINK_TRACE_OPTION_SYSGOOD)) {
			perror("pink_trace_setup");
			pink_trace_kill(pid);
			return EXIT_FAILURE;
		}

		/* The child is ready for tracing, nothing interesting in this
		 * example, let the child resume its execution. */
		pink_trace_resume(pid, 0);

		/* Wait for the child to exit. */
		waitpid(pid, &status, 0);

		return EXIT_SUCCESS;
	}
}
